### PR TITLE
Mine with multiprocessing, but with nicer stats output 

### PR DESCRIPTION
**Pull request by lk16**
This PR takes changes from @dend456 on his branch (PR #20) and adds:
* some messaging with a queue to allow nicer output with multiprocessing, just as in my nice-output PR
* changes the miner function into a class, to split functionality into smaller functions that are easier to read and maintain.

Sample output:
```
2021-05-16 17:45:28 | uptime: 0:03:43 |  45 hash/sec | mined:   0 SKEPTI |  0.00 SKEPTI/h
2021-05-16 17:45:29 | uptime: 0:03:44 |  46 hash/sec | mined:   0 SKEPTI |  0.00 SKEPTI/h
```

During loading of the chain the output is a big mess, but I don't think that should be addressed in this PR.

### CODE REVIEW

**Comment by lk16** [comment]
@sashimi-houdini I don't understand what flake8 is complaining about in CI. When I run `flake8 . --show-source --statistics` locally, it doesn't find the remaining issue. Furthermore line 222 of  `skepticoin/explorer.py` doesn't even exist on my PR. I did not even touch that file in this PR. Do you have an idea what's happening?

**Comment by sashimi-houdini** [comment]
> I don't understand what flake8 is complaining about in CI.

I do, and it's my fault. I'll fix it.

I guess github performs its actions on the merge point of your PR and master, and the problem is on master (introduced by yours truly)

File: skepticoin/scripts/mine.py [hunk 1]
```diff
@@ -23,83 +28,193 @@
 )
 
 
-def main():
-    parser = DefaultArgumentParser()
-    args = parser.parse_args()
-    configure_logging_from_args(args)
+def run_miner(args: Any, wallet_lock: synchronize.Lock, q: Queue, miner_id: int) -> None:
```
Comment (sashimi-houdini) [inline skepticoin/scripts/mine.py:31]: `q` -> `queue`, here and elsewhere (bikeshedding, I know)

File: skepticoin/scripts/mine.py [hunk 2]
```diff
@@ -23,83 +28,193 @@
...
+            os.exit(0)
+
+    def get_key_for_mined_block(self) -> bytes:
+        self.wallet_lock.acquire()
+        public_key = self.wallet.get_annotated_public_key("reserved for potentially mined block")
+        save_wallet(self.wallet)
+        self.wallet_lock.release()
+        return public_key
+
+    def get_balance(self) -> float:
```
Comment (sashimi-houdini) [inline skepticoin/scripts/mine.py:78]: `float` -> `Decimal`

File: skepticoin/scripts/mine.py [hunk 3]
```diff
@@ -23,83 +28,193 @@
...
+                block = construct_block_for_mining(
+                    self.coinstate, transactions, SECP256k1PublicKey(public_key), increasing_time, b'', nonce)
+
+                hashes += 1
+                nonce = (nonce + 1) % (1 << 32)
+
+                if block.hash() < block.target:
+                    self.handle_mined_block(block)
+
+                    # get key for next mined block
```
Comment (sashimi-houdini) [inline skepticoin/scripts/mine.py:126]: comment redundant

**Review by sashimi-houdini — COMMENTED** [review]
Other than the minor nitpicks this looks good to me!

File: skepticoin/scripts/mine.py (hunk 1 above)
Comment (lk16) [reply skepticoin/scripts/mine.py]: I had to `import queue` earlier, but that import is gone now so using `queue` works again.

**Comment by lk16** [comment]
@sashimi-houdini I addressed all your comments, so I think this is ready to be merged.

**Pull request merged**